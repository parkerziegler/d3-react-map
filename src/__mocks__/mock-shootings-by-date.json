[
  [
    {
      "x": 1420088400000,
      "y": 91
    },
    {
      "x": 1422766800000,
      "y": 83
    },
    {
      "x": 1425186000000,
      "y": 113
    },
    {
      "x": 1427860800000,
      "y": 102
    },
    {
      "x": 1430452800000,
      "y": 86
    },
    {
      "x": 1433131200000,
      "y": 80
    },
    {
      "x": 1435723200000,
      "y": 124
    },
    {
      "x": 1438401600000,
      "y": 102
    },
    {
      "x": 1441080000000,
      "y": 97
    },
    {
      "x": 1443672000000,
      "y": 91
    },
    {
      "x": 1446350400000,
      "y": 84
    },
    {
      "x": 1448946000000,
      "y": 93
    }
  ],
  [
    {
      "x": 1430452800000,
      "y": 87
    },
    {
      "x": 1420088400000,
      "y": 87
    },
    {
      "x": 1422766800000,
      "y": 100
    },
    {
      "x": 1425186000000,
      "y": 100
    },
    {
      "x": 1427860800000,
      "y": 86
    },
    {
      "x": 1433131200000,
      "y": 100
    },
    {
      "x": 1435723200000,
      "y": 89
    },
    {
      "x": 1438401600000,
      "y": 94
    },
    {
      "x": 1441080000000,
      "y": 84
    },
    {
      "x": 1443672000000,
      "y": 87
    },
    {
      "x": 1448946000000,
      "y": 85
    },
    {
      "x": 1446350400000,
      "y": 94
    }
  ]
]
