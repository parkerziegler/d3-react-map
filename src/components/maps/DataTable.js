import * as React from 'react'
import { connect } from 'react-redux';
import '../../App.css'
import * as _ from 'lodash';

class DataTable extends React.Component {

	constructor() {
		super();
		this.state = { activeTab: 'highest', isHoverOnHighest: false, isHoverOnLowest: false };
		this.onMouseEnterHandler = this.onMouseEnterHandler.bind(this);
		this.onMouseLeaveHandler = this.onMouseLeaveHandler.bind(this);
		this.onClickHandler = this.onClickHandler.bind(this);
	}

	onMouseEnterHandler(event) {

		if (event.target.id === 'highest') {
			this.setState({
				...this.state,
				isHoverOnHighest: true,
				isHoverOnLowest: false
			});
		} else if (event.target.id === 'lowest') {
			this.setState({
				...this.state,
				isHoverOnHighest: false,
				isHoverOnLowest: true
			});
		} else {
			return;
		}
	}

	onMouseLeaveHandler(event) {

			this.setState({
				...this.state,
				isHoverOnHighest: false,
				isHoverOnLowest: false
			});
	}

	onClickHandler(event) {
		if (event.target.id === 'highest') {
			this.setState({
				...this.state,
				activeTab: 'highest'
			});
		} else if (event.target.id === 'lowest') {
			this.setState({
				...this.state,
				activeTab: 'lowest'
			});
		} else {
			return;
		}
	}

	render() {

		let active = {background: '#4292C6', color: '#FFFFFF', border: '1px solid #6BAED6'};
		let inactive = {background: '#FFFFFF', color: '#4292C6', border: '1px solid #4292C6'};

		let highest = this.state.isHoverOnHighest || this.state.activeTab === 'highest' ? active : inactive;
		let lowest = this.state.isHoverOnLowest || this.state.activeTab === 'lowest' ? active : inactive;

		let ascDesc = this.state.activeTab === 'highest' ? ['desc'] : ['asc'];

		let states = this.props.maps.data.objects ?
				_.slice(_.orderBy(this.props.maps.data.objects.states.geometries, ['properties.shootingsPerCapita'], ascDesc), 0, 5) : null;

		let stateNames = _.map(states, (state, i) => {
			return <div className='row' key={i}>{state.properties.stateName}</div>;
		});

		let shootingsPerMillion = _.map(states, (state, i) => {
			return <div className='row' key={i}>{_.round(state.properties.shootingsPerCapita * 1000000, 2)}</div>;
		});

		return (
			<div className='data-table'>
				<div className='table-button-container'>
					<div className='button' id='highest' style={highest} onMouseEnter={this.onMouseEnterHandler} onMouseLeave={this.onMouseLeaveHandler} onClick={this.onClickHandler}>Highest</div>
					<div className='button' id='lowest' style={lowest} onMouseEnter={this.onMouseEnterHandler} onMouseLeave={this.onMouseLeaveHandler}
					onClick={this.onClickHandler}>Lowest</div>
				</div>
				<div className='table'>
					<div>
						<div className='table-header'>State</div>
						{stateNames}
					</div>
					<div>
						<div className='table-header'>Shootings Per Million</div>
						{shootingsPerMillion}
					</div>
				</div>
			</div>
		);
	}
}

function mapStateToProps (state, ownProps) {
  return {
    maps: state.rootReducer
  }
}

export default connect(mapStateToProps)(DataTable);
