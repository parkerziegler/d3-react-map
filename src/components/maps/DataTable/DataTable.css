/** A constants SCSS partial for storing common variables and mixins */
@import url("https://fonts.googleapis.com/css?family=Roboto");
.data-table {
  display: flex;
  flex-direction: column;
  flex: 0 0 auto; }
  .data-table .table-button-container {
    display: flex;
    flex-direction: row;
    justify-content: space-around;
    width: 60%; }
    .data-table .table-button-container .button {
      font-size: 10px;
      font-family: "HelveticaNeue", "Helvetica", sans-serif;
      width: 40%;
      height: auto;
      background: #50505E;
      flex: 0 0 auto;
      padding: 5px 5px 5px 5px;
      text-align: center;
      cursor: pointer; }
      @media screen and (min-width: 480px) {
        .data-table .table-button-container .button {
          font-size: 12px; } }
      @media screen and (min-width: 768px) {
        .data-table .table-button-container .button {
          font-size: 15px; } }
      @media screen and (min-width: 1024px) {
        .data-table .table-button-container .button {
          font-size: 18px; } }
    .data-table .table-button-container div:first-child {
      border-top-left-radius: 20px;
      border-bottom-left-radius: 20px; }
    .data-table .table-button-container div:last-child {
      border-top-right-radius: 20px;
      border-bottom-right-radius: 20px; }
  .data-table .table {
    display: flex;
    flex-direction: row;
    margin-top: 10px; }
    .data-table .table div {
      display: flex;
      flex-direction: column;
      width: 80%; }
      .data-table .table div .table-header {
        font: 700 13px "HelveticaNeue", "Helvetica", sans-serif;
        padding: 5px 0 5px 0; }
      .data-table .table div .row {
        font: 400 12px "Roboto", sans-serif;
        padding: 2px 0 2px 0; }
