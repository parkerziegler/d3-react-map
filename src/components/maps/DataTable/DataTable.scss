@import '../../../stylesheets/constants.scss';

$button-font-sizes: (
    null: 10px,
    mobile: 12px,
    tablet: 15px,
    desktop: 18px
);

.data-table {
    @include flex-column;

    .table-button-container {
        @include flex-row;
        justify-content: space-around;
        width: 60%;

        .button {
            @include font-size($button-font-sizes);
            font-family: $helvetica;
            width: 40%;
            height: auto;
            background: #50505E;
            flex: 0 0 auto;
            padding: 5px 5px 5px 5px;
            text-align: center;
            cursor: pointer;
        }

        div:first-child {
            border-top-left-radius: 20px;
            border-bottom-left-radius: 20px;
        }

        div:last-child {
            border-top-right-radius: 20px;
            border-bottom-right-radius: 20px;
        }
    }

    .table {
        @include flex-row;
        margin-top: 10px;

        div {
            @include flex-column;
            width: 80%;

            .table-header {
                font: 700 13px $helvetica;
                padding: 5px 0 5px 0;
            }

            .row {
                font: 400 12px $roboto;
                padding: 2px 0 2px 0;
            }
        }
    }
}