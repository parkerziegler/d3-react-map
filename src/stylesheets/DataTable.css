/** A constants SCSS partial for storing common variables and mixins */
@import url("https://fonts.googleapis.com/css?family=Roboto");
.data-table {
  display: flex;
  flex-direction: column;
  flex: 0 0 auto; }
  .data-table .table-button-container {
    display: flex;
    flex-direction: row;
    justify-content: space-around; }
    .data-table .table-button-container .button {
      font-size: 10px;
      flex: 1;
      font-family: "HelveticaNeue", "Helvetica", sans-serif;
      background: #50505e;
      padding: 10px;
      text-align: center;
      cursor: pointer;
      outline: 0;
      border: 0; }
      @media screen and (min-width: 480px) {
        .data-table .table-button-container .button {
          font-size: 12px; } }
      @media screen and (min-width: 768px) {
        .data-table .table-button-container .button {
          font-size: 14px; } }
      @media screen and (min-width: 1024px) {
        .data-table .table-button-container .button {
          font-size: 16px; } }
      .data-table .table-button-container .button:first-child {
        border-top-left-radius: 10px;
        border-bottom-left-radius: 10px; }
      .data-table .table-button-container .button:last-child {
        border-top-right-radius: 10px;
        border-bottom-right-radius: 10px; }
  .data-table .table {
    display: flex;
    flex-direction: row;
    margin-top: 10px; }
    .data-table .table div {
      display: flex;
      flex-direction: column;
      width: 80%; }
      .data-table .table div .table-header {
        font: 700 13px "HelveticaNeue", "Helvetica", sans-serif;
        padding: 5px 0 5px 0; }
      .data-table .table div .row {
        font: 400 12px "Roboto", sans-serif;
        padding: 2px 0 2px 0; }
