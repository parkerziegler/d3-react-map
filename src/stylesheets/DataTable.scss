@import './_constants.scss';

$button-font-sizes: (
  null: 10px,
  mobile: 12px,
  tablet: 14px,
  desktop: 16px,
);

.data-table {
  @include flex-column;
  flex: 0 0 auto;

  .table-button-container {
    @include flex-row;
    justify-content: space-around;

    .button {
      @include font-size($button-font-sizes);
      flex: 1;
      font-family: $helvetica;
      background: #50505e;
      padding: 10px;
      text-align: center;
      cursor: pointer;
      outline: 0;
      border: 0;

      &:first-child {
        border-top-left-radius: 10px;
        border-bottom-left-radius: 10px;
      }

      &:last-child {
        border-top-right-radius: 10px;
        border-bottom-right-radius: 10px;
      }
    }
  }

  .table {
    @include flex-row;
    margin-top: 10px;

    div {
      @include flex-column;
      width: 80%;

      .table-header {
        font: 700 13px $helvetica;
        padding: 5px 0 5px 0;
      }

      .row {
        font: 400 12px $roboto;
        padding: 2px 0 2px 0;
      }
    }
  }
}
