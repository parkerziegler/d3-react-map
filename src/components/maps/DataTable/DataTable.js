import * as React from 'react';
import { connect } from 'react-redux';
import '../../../stylesheets/DataTable.css';
import { round, orderBy } from 'lodash';

class DataTable extends React.Component {
  constructor() {
    super();
    this.state = {
      activeTab: 'highest',
      isHoverOnHighest: false,
      isHoverOnLowest: false,
    };
    this.onMouseEnterHandler = this.onMouseEnterHandler.bind(this);
    this.onMouseLeaveHandler = this.onMouseLeaveHandler.bind(this);
    this.onClickHandler = this.onClickHandler.bind(this);
    this.getTableNodes = this.getTableNodes.bind(this);

    // a lookup to store names for column headers
    this.styles = {
      choropleth: {
        columnHeader: 'Shootings Per Million',
        active: {
          background: '#4292C6',
          color: '#FFFFFF',
          border: '1px solid #6BAED6',
        },
        inactive: {
          background: '#FFFFFF',
          color: '#4292C6',
          border: '1px solid #4292C6',
        },
      },
      proportional: {
        columnHeader: 'Total Shootings',
        active: {
          background: '#B24739',
          color: '#FFFFFF',
          border: '1px solid 7F3329',
        },
        inactive: {
          background: '#FFFFFF',
          color: '#B24739',
          border: '1px solid #B24739',
        },
      },
    };
  }

  onMouseEnterHandler(event) {
    if (event.target.id === 'highest') {
      this.setState({
        isHoverOnHighest: true,
        isHoverOnLowest: false,
      });
    } else if (event.target.id === 'lowest') {
      this.setState({
        isHoverOnHighest: false,
        isHoverOnLowest: true,
      });
    }
  }

  onMouseLeaveHandler(event) {
    this.setState({
      isHoverOnHighest: false,
      isHoverOnLowest: false,
    });
  }

  onClickHandler(event) {
    this.setState({
      activeTab: event.target.id,
    });
  }

  getTableNodes() {
    const { mapType, maps } = this.props;
    const { activeTab } = this.state;

    // determine whether we should render the table
    // in ascending or descending order
    // hand this off to lodash
    const ascDesc = activeTab === 'highest' ? ['desc'] : ['asc'];

    if (mapType === 'choropleth') {
      // order states by per million count
      const states = orderBy(
        maps.geoData.objects.states.geometries,
        state =>
          state.properties.numShootings / state.properties.population * 1000000,
        ascDesc
      ).slice(0, 5);

      // render divs for both the state name and
      // shootings per capita
      const stateNames = states.map((state, i) => (
        <div className="row" key={i}>
          {state.properties.stateName}
        </div>
      ));

      const stateStats = states.map((state, i) => (
        <div className="row" key={i}>
          {round(
            state.properties.numShootings /
              state.properties.population *
              1000000,
            2
          )}
        </div>
      ));

      // return these in an object to be rendered
      return {
        stateNames,
        stateStats,
      };
    } else if (mapType === 'proportional') {
      // obtain our table data by looking at our reducer
      const states = orderBy(
        maps.geoData.objects.states.geometries,
        ['properties.numShootings'],
        ascDesc
      ).slice(0, 5);

      // next render divs for both the state name and
      // total number of shootings
      const stateNames = states.map((state, i) => (
        <div className="row" key={i}>
          {state.properties.stateName}
        </div>
      ));

      const stateStats = states.map((state, i) => (
        <div className="row" key={i}>
          {state.properties.numShootings}
        </div>
      ));

      // return these in an object to be rendered
      return {
        stateNames,
        stateStats,
      };
    }
  }

  render() {
    const { stateNames, stateStats } = this.getTableNodes();
    const highest =
      this.state.isHoverOnHighest || this.state.activeTab === 'highest'
        ? this.styles[this.props.mapType].active
        : this.styles[this.props.mapType].inactive;

    const lowest =
      this.state.isHoverOnLowest || this.state.activeTab === 'lowest'
        ? this.styles[this.props.mapType].active
        : this.styles[this.props.mapType].inactive;

    return (
      <div className="data-table">
        <div className="table-button-container">
          <div
            className="button"
            id="highest"
            style={highest}
            onMouseEnter={this.onMouseEnterHandler}
            onMouseLeave={this.onMouseLeaveHandler}
            onClick={this.onClickHandler}
          >
            Highest
          </div>
          <div
            className="button"
            id="lowest"
            style={lowest}
            onMouseEnter={this.onMouseEnterHandler}
            onMouseLeave={this.onMouseLeaveHandler}
            onClick={this.onClickHandler}
          >
            Lowest
          </div>
        </div>
        <div className="table">
          <div>
            <div className="table-header">State</div>
            {stateNames}
          </div>
          <div>
            <div className="table-header">
              {this.styles[this.props.mapType].columnHeader}
            </div>
            {stateStats}
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  maps: state.mapReducer,
});

export default connect(mapStateToProps)(DataTable);
