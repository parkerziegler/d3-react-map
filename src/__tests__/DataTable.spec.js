import DataTable from '../components/maps/DataTable/DataTable';
import { mockStore } from '../__mocks__/mock-store';

describe('<DataTable /> component', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <DataTable mapType="choropleth" store={mockStore} />
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });

  it('updates local state on button hover and click', () => {
    const wrapper = shallow(
      <DataTable mapType="choropleth" store={mockStore} />
    ).dive();
    const node = wrapper.find('#lowest');
    node.simulate('mouseenter', { target: { id: 'lowest' } });
    expect(wrapper.state('isHoverOnLowest')).toBe(true);
    expect(wrapper.state('isHoverOnHighest')).toBe(false);

    node.simulate('click', { target: { id: 'lowest' } });
    expect(wrapper.state('activeTab')).toEqual('lowest');
    node.simulate('mouseleave');
    expect(wrapper.state('isHoverOnLowest')).toBe(false);
  });
});
