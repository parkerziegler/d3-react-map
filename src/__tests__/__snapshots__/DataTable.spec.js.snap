// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataTable /> component renders correctly 1`] = `
<div
  className="data-table"
>
  <div
    className="table-button-container"
  >
    <div
      className="button"
      id="highest"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "background": "#4292C6",
          "border": "1px solid #6BAED6",
          "color": "#FFFFFF",
        }
      }
    >
      Highest
    </div>
    <div
      className="button"
      id="lowest"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "background": "#FFFFFF",
          "border": "1px solid #4292C6",
          "color": "#4292C6",
        }
      }
    >
      Lowest
    </div>
  </div>
  <div
    className="table"
  >
    <div>
      <div
        className="table-header"
      >
        State
      </div>
      <div
        className="row"
        key="0"
      >
        New Mexico
      </div>
      <div
        className="row"
        key="1"
      >
        District of Columbia
      </div>
      <div
        className="row"
        key="2"
      >
        Oklahoma
      </div>
      <div
        className="row"
        key="3"
      >
        Alaska
      </div>
      <div
        className="row"
        key="4"
      >
        Arizona
      </div>
    </div>
    <div>
      <div
        className="table-header"
      >
        Shootings Per Million
      </div>
      <div
        className="row"
        key="0"
      >
        20.63
      </div>
      <div
        className="row"
        key="1"
      >
        18.53
      </div>
      <div
        className="row"
        key="2"
      >
        17.92
      </div>
      <div
        className="row"
        key="3"
      >
        17.73
      </div>
      <div
        className="row"
        key="4"
      >
        14.15
      </div>
    </div>
  </div>
</div>
`;
